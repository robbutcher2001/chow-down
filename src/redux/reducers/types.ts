export interface Message {
    user: string
    message: string
}

export interface HelloState {
    hello_messages: Message[]
}

export const SAY_HELLO = 'SAY_HELLO'
export const DELETE_HELLO = 'DELETE_HELLO';

interface SayHelloAction {
    type: typeof SAY_HELLO
    rob_payload: Message
}

interface DeleteHelloAction {
    type: typeof DELETE_HELLO
}

export type SayHelloTypes = SayHelloAction | DeleteHelloAction