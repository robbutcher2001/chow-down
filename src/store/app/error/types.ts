import { Action } from 'redux';

export enum ErrorActionTypes {
  UNEXPECTED_SERVER_ERROR = '@@common/UNEXPECTED_SERVER_ERROR',
  UNEXPECTED_RESPONSE = '@@common/UNEXPECTED_RESPONSE',
  CLEAR_ERROR = '@@common/CLEAR_ERROR'
}

export interface ErrorState {
  readonly message: string
}

export interface ErrorMessageApiResponse {
  message: string
}

export interface ErrorApiResponse extends Action {
  type: ErrorActionTypes,
  json: ErrorMessageApiResponse
}