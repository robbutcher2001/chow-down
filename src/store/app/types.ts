import { Action } from 'redux';

export enum AppActionTypes {
    UNEXPECTED_SERVER_ERROR = '@@common/UNEXPECTED_SERVER_ERROR',
    UNEXPECTED_RESPONSE = '@@common/UNEXPECTED_RESPONSE',
    CLEAR_ERROR = '@@common/CLEAR_ERROR'
}

export interface AppState {
    readonly error: string
}

export interface ErrorMessageApiResponse {
    message: string
}

export interface ErrorApiResponse extends Action {
    type: AppActionTypes,
    json: ErrorMessageApiResponse
}