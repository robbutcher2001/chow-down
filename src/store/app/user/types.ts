import { Action } from 'redux';

export enum UserActionTypes {
  SET_USER_IS_SELECTING_DAY = '@@common/SET_USER_IS_SELECTING_DAY',
  CLEAR_USER_IS_SELECTING_DAY = '@@common/CLEAR_USER_IS_SELECTING_DAY'
}

export interface UserState {
  readonly selectedDay: string
}

export interface UserAction extends Action {
  type: UserActionTypes,
  day: string
}