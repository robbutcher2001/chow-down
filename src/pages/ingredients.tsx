import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { GlobalState } from '../store';
import { Ingredient, GetIngredientsApiRequest, PostIngredientApiRequest } from '../store/domain/ingredients/types';
import { getIngredientsRequest, postIngredientsRequest } from '../store/domain/ingredients/actions';

import Form from '../components/form';
import InputBox from '../components/input-box';

interface StateProps {
    error: string,
    failure: string,
    ingredients: Ingredient[],
    ui: {
        pending: {
            get: boolean,
            post: boolean
        }
    }
};

interface DispatchProps {
    getIngredients: () => GetIngredientsApiRequest,
    postIngredient: (form: object) => PostIngredientApiRequest
};

interface OwnProps { };

interface OwnState { };

type CombinedProps = StateProps & DispatchProps & OwnProps;

class IngredientsPage extends Component<CombinedProps, OwnState> {
    constructor(props: CombinedProps) {
        super(props);
    }

    requestIngredients = () => this.props.getIngredients();
    addIngredient = (form: object) => this.props.postIngredient(form);

    componentDidMount = () => this.props.getIngredients();

    render = () => {
        console.log(this.props.ui.pending.post);
        return (
            <div>
                <h4>List ingredients</h4>
                <button onClick={this.requestIngredients}>
                    Refresh ingredients
                </button>
                {this.props.ui.pending.get &&
                    <div style={{ color: 'red' }}>Getting..</div>
                }
                <ul>
                    {this.props.ingredients.map((ingredient, index) =>
                        <li key={index}>{ingredient.name}</li>
                    )}
                </ul>
                <h4>Add ingredient</h4>
                <Form
                    dispatch={this.addIngredient}
                    submitText='Add ingredient'>
                    <InputBox
                        name='name'
                        placeholderText='New ingredient name'
                    />
                </Form>
                <div>{this.props.error}</div>
                {this.props.ui.pending.post &&
                    <div>Adding your new ingredient..</div>
                }
                {this.props.failure &&
                    <div style={{color: 'red'}}>{this.props.failure}</div>
                }
            </div>
        );
    }
};

const mapStateToProps = ({ app, domain, ui }: GlobalState, ownProps: OwnProps): StateProps => ({
    //TODO: move application-wide errors to footer component for toast notification (feed down through page container)
    error: app.error,
    failure: domain.ingredient.failure,
    ingredients: domain.ingredient.ingredients,
    ui: {
        pending: {
            get: ui.ingredient.getPending,
            post: ui.ingredient.postPending
        }
    }
});

const mapDispatchToProps = (dispatch: Dispatch, ownProps: OwnProps): DispatchProps => ({
    getIngredients: () => dispatch(getIngredientsRequest()),
    postIngredient: (form: object) => dispatch(postIngredientsRequest(form))
});

export default connect<StateProps, DispatchProps, OwnProps, GlobalState>
    (mapStateToProps, mapDispatchToProps)(IngredientsPage);