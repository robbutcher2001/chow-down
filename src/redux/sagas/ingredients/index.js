'use strict';

import { all, takeLatest } from 'redux-saga/effects';

import getSaga from './get';
import postSaga from './post';

export default function* watcherSaga() {
    console.log('[saga] Started ingredients saga');
    yield all([
        takeLatest('GET_INGREDIENTS_REQUEST', getSaga),
        takeLatest('POST_INGREDIENT_REQUEST', postSaga)
    ]);
}