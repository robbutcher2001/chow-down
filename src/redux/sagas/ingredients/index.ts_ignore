'use strict';

import { all, takeLatest } from 'redux-saga/effects';

import { Actions } from '../../../globals/constants';

import getSaga from './get';
import postSaga from './post';

export default function* watcherSaga() {
    console.log('[saga] Started ingredients saga');
    yield all([
        takeLatest(Actions.ingredients.GET_INGREDIENTS_REQUEST, getSaga),
        takeLatest(Actions.ingredients.POST_INGREDIENT_REQUEST, postSaga)
    ]);
}