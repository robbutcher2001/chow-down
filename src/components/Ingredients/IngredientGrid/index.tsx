import React, { ReactNode } from 'react';

import styled from 'styled-components';

import { Ingredient } from '../../../store/domain/ingredients/types';
import IngredientCard from '../IngredientCard';
import IngredientMarker from '../IngredientMarker';
import { NegativeBox } from '../../MessageBox';

interface IngredientGridProps {
  isLoading: boolean,
  title: string,
  ingredients: Ingredient[]
};

const IngredientGrid = styled.div`
  h3 {
    color: ${props => props.theme.isDark ?
      props.theme.colour.lightestGrey :
      props.theme.colour.black
    };
  }

  ul {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    list-style-type: none;
    margin: 0;
    padding: 0;
    min-height: 120px;

    li:first-child {
      margin-top: 0.5rem;
    }
  }
`

const ingredientsWithMarkers = (ingredients: Ingredient[]): ReactNode[] => {
  let previousLetter: string;

  return ingredients.reduce((arr, ingredient, i) => {
    if (ingredient.name) {
      const currentLetter: string = ingredient.name.charAt(0).toLocaleLowerCase();

      if (currentLetter !== previousLetter) {
        previousLetter = currentLetter;
        arr.push(<IngredientMarker
          key={currentLetter}
          mark={currentLetter.toLocaleUpperCase()}
        />);
      }

      arr.push(<IngredientCard
        key={i}
        ingredient={ingredient}
      />);
    }

    return arr;
  }, [] as ReactNode[]);
};

export default (props: IngredientGridProps) =>
  !props.isLoading && props.ingredients.length === 0 ?
    <NegativeBox message='No ingredients yet!' /> :
    <IngredientGrid>
      <h3>{props.title}</h3>
      <ul className={props.isLoading ? 'spinner spinning' : 'spinner'}>
        {ingredientsWithMarkers(props.ingredients)}
      </ul>
    </IngredientGrid>;