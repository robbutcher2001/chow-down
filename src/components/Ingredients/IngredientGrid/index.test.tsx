import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import IngredientGrid from '.';
import { Ingredient } from '../../../store/domain/ingredients/types';

const mockIngredients: Ingredient[] = [{
  id: '123',
  ingredient: 'fake_ingredient1'
},
{
  id: '124',
  ingredient: 'fake_ingredient2'
}];

test('IngredientGrid basic snapshot render', () => {
  const ingredientGrid = renderer.create(
    <IngredientGrid ingredients={mockIngredients} />
  );

  expect(ingredientGrid.toJSON()).toMatchSnapshot();
});

test('IngredientGrid correct tag content assertion', () => {
  const { getByText } = render(
    <IngredientGrid ingredients={mockIngredients} />
  );

  expect(getByText(/fake_ingredient1/).textContent).toEqual('fake_ingredient1');
});

test('IngredientGrid render multiple ingredients assertion', () => {
  const { getByText } = render(
    <IngredientGrid ingredients={mockIngredients} />
  );

  expect(getByText(/fake_ingredient1/).textContent).toEqual('fake_ingredient1');
  expect(getByText(/fake_ingredient2/).textContent).toEqual('fake_ingredient2');
});