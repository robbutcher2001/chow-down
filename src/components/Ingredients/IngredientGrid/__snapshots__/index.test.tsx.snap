// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IngredientGrid basic isLoading snapshot render 1`] = `
<ul
  className="sc-AxhCb bjixBq spinner spinning"
>
  <li
    className="sc-AxirZ zzNUa"
  >
    F
  </li>
  <li
    className="sc-AxjAm kiPfQz"
  >
    fake_ingredient1
  </li>
  <li
    className="sc-AxjAm kiPfQz"
  >
    fake_ingredient2
  </li>
</ul>
`;

exports[`IngredientGrid basic snapshot render 1`] = `
<ul
  className="sc-AxhCb bjixBq spinner"
>
  <li
    className="sc-AxirZ zzNUa"
  >
    F
  </li>
  <li
    className="sc-AxjAm kiPfQz"
  >
    fake_ingredient1
  </li>
  <li
    className="sc-AxjAm kiPfQz"
  >
    fake_ingredient2
  </li>
</ul>
`;
